File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/gpx/GpxDynLayer.java
package ch.bailu.aat_lib.map.layer.gpx;

import ch.bailu.aat_lib.preferences.map.SolidLegend;
import ch.bailu.aat_lib.dispatcher.DispatcherInterface;
import ch.bailu.aat_lib.dispatcher.OnContentUpdatedInterface;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.GpxInformationCache;
import ch.bailu.aat_lib.gpx.interfaces.GpxType;
import ch.bailu.aat_lib.map.MapContext;
import ch.bailu.aat_lib.util.Point;
import ch.bailu.aat_lib.map.layer.MapLayerInterface;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.service.ServicesInterface;

public final class GpxDynLayer implements MapLayerInterface, OnContentUpdatedInterface {
    private final GpxInformationCache infoCache = new GpxInformationCache();

    private GpxLayer gpxOverlay;
    private GpxLayer legendOverlay;

    private final SolidLegend slegend;

    private final MapContext mcontext;

    private final ServicesInterface services;

    public GpxDynLayer(StorageInterface s, MapContext mc, ServicesInterface services) {
        mcontext = mc;
        this.services = services;
        slegend = new SolidLegend(s, mcontext.getSolidKey());

        createLegendOverlay();
        createGpxOverlay();
    }

    public GpxDynLayer(StorageInterface s, MapContext mc, ServicesInterface services,
                         DispatcherInterface dispatcher, int iid) {
        this(s, mc, services);
        dispatcher.addTarget(this, iid);
    }

    @Override
    public void drawInside(MapContext mcontext) {
        gpxOverlay.drawInside(mcontext);
        legendOverlay.drawInside(mcontext);
    }

    @Override
    public void drawForeground(MapContext mcontext) {}

    @Override
    public boolean onTap(Point tapPos) {
        return false;
    }

    private GpxType type = GpxType.NONE;

    @Override
    public void onContentUpdated(int iid, GpxInformation info) {
        infoCache.set(iid, info);


        if (type != toType(info)) {
            type = toType(info);
            createGpxOverlay();
            createLegendOverlay();
        }

        infoCache.letUpdate(gpxOverlay);
        infoCache.letUpdate(legendOverlay);

        mcontext.getMapView().requestRedraw();
    }



    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {

        if (slegend.hasKey(key)) {
            createLegendOverlay();
            infoCache.letUpdate(legendOverlay);
            mcontext.getMapView().requestRedraw();
        }
    }

    private static GpxType toType(GpxInformation i) {
        if (i != null && i.getGpxList() != null) {
            return i.getGpxList().getDelta().getType();
        }
        return GpxType.NONE;
    }


    private void createGpxOverlay() {
        GpxType type = toType(infoCache.info);
        gpxOverlay = Factory.get(type).layer(mcontext, services,0);
    }


    private void createLegendOverlay() {
        GpxType type = toType(infoCache.info);
        legendOverlay = Factory.get(type).legend(slegend.getStorage(), slegend, 0);
    }


    @Override
    public void onLayout(boolean changed, int l, int t, int r, int b) {}

    @Override
    public void onAttached() {}

    @Override
    public void onDetached() {}
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/gpx/GpxDynLayer.kt
package ch.bailu.aat_lib.map.layer.gpx

import ch.bailu.aat_lib.dispatcher.DispatcherInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.gpx.information.GpxInformationCache
import ch.bailu.aat_lib.gpx.interfaces.GpxType
import ch.bailu.aat_lib.map.MapContext
import ch.bailu.aat_lib.map.layer.MapLayerInterface
import ch.bailu.aat_lib.map.layer.gpx.legend.GpxLegendLayer
import ch.bailu.aat_lib.map.layer.gpx.legend.NullLegendWalker
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.map.SolidLayerType
import ch.bailu.aat_lib.preferences.map.SolidLegend
import ch.bailu.aat_lib.service.ServicesInterface
import ch.bailu.aat_lib.util.Point

class GpxDynLayer(
    storage: StorageInterface,
    private val mcontext: MapContext,
    private val services: ServicesInterface
) : MapLayerInterface, TargetInterface {
    private val infoCache = GpxInformationCache()
    private var gpxOverlay: GpxLayer = TrackLayer(mcontext)
    private var legendOverlay: GpxLayer = GpxLegendLayer(NullLegendWalker())
    private val slegend: SolidLegend = SolidLegend(storage, mcontext.getSolidKey())
    private val solidLayerType = SolidLayerType(storage)
    private var forceReconfigure = true
    private var type = GpxType.TRACK

    constructor(
        storage: StorageInterface, mc: MapContext, services: ServicesInterface,
        dispatcher: DispatcherInterface, iid: Int
    ) : this(storage, mc, services) {
        dispatcher.addTarget(this, iid)
    }

    override fun drawInside(mcontext: MapContext) {
        gpxOverlay.drawInside(mcontext)
        legendOverlay.drawInside(mcontext)
    }
    override fun drawForeground(mcontext: MapContext) {}

    override fun onTap(tapPos: Point): Boolean {
        return false
    }

    override fun onContentUpdated(iid: Int, info: GpxInformation) {
        infoCache.set(iid, info)
        if (forceReconfigure || type !== info.getType()) {
            forceReconfigure = false
            type = info.getType()
            createGpxOverlay(iid)
            createLegendOverlay()
        }
        infoCache.letUpdate(gpxOverlay)
        infoCache.letUpdate(legendOverlay)
        mcontext.getMapView().requestRedraw()
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (slegend.hasKey(key) || solidLayerType.hasKey(key)) {
            forceReconfigure = true
            infoCache.letUpdate(this)
        }
    }

    private fun createGpxOverlay(iid: Int) {
        val type = infoCache.info.getType()
        gpxOverlay = Factory[type].layer(mcontext, services, solidLayerType, iid)
    }

    private fun createLegendOverlay() {
        val type = infoCache.info.getType()
        legendOverlay = Factory[type].legend(slegend)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}
    override fun onAttached() {
        // If coming back from the preferences dialog
        forceReconfigure = true
        infoCache.letUpdate(this)
    }

    override fun onDetached() {}
}
